File: packages/llm/src/ollama.ts
import type { AppSettings } from "@localrag/config";
import type { ChatMessage, LlmCompleteResult, LlmProvider } from "./types.js";
import { withRetries } from "./retry.js";

function joinUrl(base: string, path: string) {
  return `${base.replace(/\/$/, "")}${path}`;
}

export async function ollamaListModels(host: string): Promise<string[]> {
  const res = await fetch(joinUrl(host, "/api/tags"));
  if (!res.ok) throw new Error(`Ollama list models failed: ${res.status}`);
  const data = (await res.json()) as { models?: { name?: string }[] };
  return (data.models ?? []).map((m) => m.name ?? "").filter(Boolean);
}

export async function ollamaEmbed(
  host: string,
  model: string,
  input: string,
): Promise<number[]> {
  return withRetries(
    async () => {
      const res = await fetch(joinUrl(host, "/api/embeddings"), {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({ model, prompt: input }),
      });
      if (!res.ok) {
        const t = await res.text();
        throw new Error(`Ollama embeddings failed: ${res.status} ${t}`);
      }
      const data = (await res.json()) as { embedding?: number[] };
      if (!data.embedding?.length) throw new Error("Ollama returned empty embedding");
      return data.embedding;
    },
    { attempts: 4, baseMs: 200, maxMs: 5000 },
  );
}

export function createOllamaLlmProvider(
  host: string,
  model: string,
): LlmProvider {
  return {
    async complete(messages: ChatMessage[]): Promise<LlmCompleteResult> {
      return withRetries(
        async () => {
          const res = await fetch(joinUrl(host, "/api/chat"), {
            method: "POST",
            headers: { "content-type": "application/json" },
            body: JSON.stringify({
              model,
              messages,
              stream: false,
            }),
          });
          if (!res.ok) {
            const t = await res.text();
            throw new Error(`Ollama chat failed: ${res.status} ${t}`);
          }
          const data = (await res.json()) as {
            message?: { content?: string };
          };
          const text = data.message?.content ?? "";
          return { text };
        },
        { attempts: 4, baseMs: 200, maxMs: 5000 },
      );
    },
  };
}

export function ollamaEmbeddingModel(settings: AppSettings): string {
  return settings.ollama.embedModel;
}

File: rag-service/rag_service/llm/ollama.py
from __future__ import annotations

from typing import Any

import httpx

from rag_service.llm.retry import with_retries


def join_url(base: str, path: str) -> str:
    return base.rstrip("/") + path


def ollama_list_models(host: str) -> list[str]:
    with httpx.Client(timeout=30.0) as client:
        res = client.get(join_url(host, "/api/tags"))
        if res.status_code != 200:
            raise RuntimeError(f"Ollama list models failed: {res.status_code}")
        data = res.json()
        models = data.get("models") or []
        return [str(m.get("name") or "") for m in models if m.get("name")]


def ollama_embed(host: str, model: str, prompt: str) -> list[float]:
    def call() -> list[float]:
        with httpx.Client(timeout=120.0) as client:
            res = client.post(
                join_url(host, "/api/embeddings"),
                json={"model": model, "prompt": prompt},
                headers={"content-type": "application/json"},
            )
            if res.status_code != 200:
                hint = (
                    f" If the model is missing, run: ollama pull {model}"
                    if res.status_code == 404
                    else ""
                )
                raise RuntimeError(
                    f"Ollama embeddings failed: {res.status_code} {res.text}{hint}"
                )
            data: dict[str, Any] = res.json()
            emb = data.get("embedding")
            if not emb or not isinstance(emb, list):
                raise RuntimeError("Ollama returned empty embedding")
            return [float(x) for x in emb]

    return with_retries(call, attempts=4, base_ms=200, max_ms=5000)


def ollama_chat(host: str, model: str, messages: list[dict[str, str]]) -> str:
    def call() -> str:
        with httpx.Client(timeout=300.0) as client:
            res = client.post(
                join_url(host, "/api/chat"),
                json={"model": model, "messages": messages, "stream": False},
                headers={"content-type": "application/json"},
            )
            if res.status_code != 200:
                hint = (
                    f" If the model is missing, run: ollama pull {model}"
                    if res.status_code == 404
                    else ""
                )
                raise RuntimeError(
                    f"Ollama chat failed: {res.status_code} {res.text}{hint}"
                )
            data = res.json()
            msg = data.get("message") or {}
            return str(msg.get("content") or "")

    return with_retries(call, attempts=4, base_ms=200, max_ms=5000)
